### PR TITLE
Support Mocking in OData Client.

### PR DESCRIPTION
**Pull request by ElizabethOkerio**
### Issues

*This pull request fixes issue #1250.*

### Description
The DataServiceContext, DataServiceQuery, DataServiceQueryOfT and DataServiceQuerySingleOfT classes cannot be mocked as they are. In this PR, I have made these classes' public methods and properties virtual to support mocking. 
Virtual methods can be mocked using any Mock Framework. Example, using the Moq Framework
You can mock the <code>AddObject() </code> method in the <code> Microsoft.OData.Client.DataServiceContext </code> class like below: 

<pre>
using Moq; 
using Xunit
namespace SampleMock
{
   /// How to mock the DataServiceContext class methods using moq framework example. 
    public class DataServiceContextMockTests 
    { 
        [Fact] 
        public void VerifyIfDataServiceContextAddObjectMethodIsMockable() 
        { 
            string str = "user"; 
            Mock<DataServiceContext> mockObject = new Mock<DataServiceContext>(); 
            mockObject.CallBase = true; 
            mockObject.Setup(x => x.AddObject(It.IsAny<string>(), It.IsAny<object>())).Callback(() => { str = "Saved Successfully"; }); 
            Users users = new Users(); 
            string result = users.SaveUser(mockObject.Object); 
            Assert.Equal(str, result); 
        }  
        /// A sample implementation class with methods calling the DataserviceContext.cs methods 
        public class Users 
        { 
            /// A simple method to save a user's details  
            /// A concrete class in the Microsoft.OData.Client library with implemented methods that this method calls.         
            public string SaveUser(DataServiceContext dataServiceContext) 
            { 
                string str = "user"; 
                object obj = null; 
                dataServiceContext.AddObject(str, obj); 
                return "Saved Successfully"; 
            } 
        } 
    }
</pre>
### Checklist (Uncheck if it is not completed)
- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: test/FunctionalTests/Microsoft.OData.Client.Tests/DataServiceContextMockTests.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        public class Users
+        {
+            /// <summary>
+            /// A simple method to save a user's details
+            /// </summary>
+            /// <param name="DataServiceContext">
+            /// A concrete class in the Microsoft.OData.Client library with implemented methods that this method calls.        
+            /// </param>
+            /// <returns>
+            /// return "Saved Successfully if the AddObject method runs successfully".
```
Comment (paulodero) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/DataServiceContextMockTests.cs]:
```suggestion
            /// return "Saved Successfully" if the AddObject method runs successfully.
```

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj [hunk 2]
```diff
@@ -15,6 +15,7 @@
     <PackageReference Include="FluentAssertions" Version="4.19.2" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.0.0" />
     <PackageReference Include="Microsoft.NETCore.Portable.Compatibility" Version="1.0.1" />
+    <PackageReference Include="Moq" Version="4.10.1" />
```
Comment (paulodero) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj]: Moq version should be the latest `4.13.1` just like the one you have used in `Microsoft.OData.Client.Tests` project I believe. And again, do we have any test case added in this project that requires `Moq`?

Comment (ElizabethOkerio) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj]: Yes. the lastest Moq version(4.13.1) supports netcoreapp2.0 but this project targets netcoreapp1.0 and this is the most up to date Moq version that is compatible with netcoreapp1.0

Comment (ElizabethOkerio) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj]: Yes there is a sample test to shoe how these methods can be mocked.

Comment (paulodero) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj]: In that case is it good to consistently use version `4.10.1` in all projects? I can only see tests added for `Microsoft.OData.Client.Tests` project and not `Microsoft.OData.Client.Tests.Netcore` project. This brings the question if we need `Moq `in `Microsoft.OData.Client.Tests.Netcore` project.

Comment (ElizabethOkerio) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj]: Basically for the test to run in a .netcore application, then the <code>Moq framework </code> 
is needed in <code>Microsoft.OData.Clieny.Tests.Netcore</code>

Comment (ElizabethOkerio) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj]: I don't see why we cannot use the latest <code>Moq Framework </code> in <code>
Microsoft.OData.Client.Tests</code> if it is supported.

**Comment by weitzhandler** [comment]
Related: https://github.com/OData/ODataConnectedService/issues/93

**Pull request merged**